staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/admission.go

```diff
@@ -283,21 +288,52 @@ func (a *GenericAdmissionWebhook) callHook(ctx context.Context, h *v1alpha1.Exte
 }
 
 func (a *GenericAdmissionWebhook) hookClient(h *v1alpha1.ExternalAdmissionHook) (*rest.RESTClient, error) {
-	serverName := h.ClientConfig.Service.Name + "." + h.ClientConfig.Service.Namespace + ".svc"
-	u, err := a.serviceResolver.ResolveEndpoint(h.ClientConfig.Service.Namespace, h.ClientConfig.Service.Name)
+	if svc := h.ClientConfig.Service; svc != nil {
+		serverName := svc.Name + "." + svc.Namespace + ".svc"
+		u, err := a.serviceResolver.ResolveEndpoint(svc.Namespace, svc.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: cache these instead of constructing one each time
+		restConfig, err := a.authInfoResolver.ClientConfigFor(serverName)
+		if err != nil {
+			return nil, err
+		}
+		cfg := rest.CopyConfig(restConfig)
+		cfg.Host = u.Host
+		if svc.Path != nil {
+			cfg.APIPath = path.Join(u.Path, *svc.Path)
+		} else {
+			cfg.APIPath = u.Path
+		}
+		cfg.TLSClientConfig.ServerName = serverName
+		cfg.TLSClientConfig.CAData = h.ClientConfig.CABundle
+		cfg.ContentConfig.NegotiatedSerializer = a.negotiatedSerializer
+		cfg.ContentConfig.ContentType = runtime.ContentTypeJSON
+		return rest.UnversionedRESTClientFor(cfg)
+	}
+
+	if h.ClientConfig.URL == nil {
+		return nil, &ErrCallingWebhook{WebhookName: h.Name, Reason: ErrNeedServiceOrURL}
+	}
+
+	u, err := url.Parse(*h.ClientConfig.URL)
 	if err != nil {
-		return nil, err
+		return nil, &ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("Unparsable URL: %v", err)}
 	}
 
 	// TODO: cache these instead of constructing one each time
-	restConfig, err := a.authInfoResolver.ClientConfigFor(serverName)
+	restConfig, err := a.authInfoResolver.ClientConfigFor(u.Host)
```

Comment:
Doesn't host include port?  I wonder what the clientConfigFor code does in that case.

Reply:
I am not sure if I should split here or let authInfoResolver worry about it-- what if people run two different things on two ports on the same host? (parameter name is "server" which is pretty ambiguous as to whether it should include a port)

Reply:
Previously we mandated 443, so at least this doesn't make it worse.